// @storybook-vue/nuxt@0.1.7/dist/preset.cjs
'use strict';

const node_path = require('node:path');
const node_url = require('node:url');
const node_module = require('node:module');
const vite = require('vite');

var _documentCurrentScript = typeof document !== 'undefined' ? document.currentScript : null;
const packageDir = node_path.resolve(node_url.fileURLToPath(
  (typeof document === 'undefined' ? require('u' + 'rl').pathToFileURL(__filename).href : (_documentCurrentScript && _documentCurrentScript.src || new URL('preset.cjs', document.baseURI).href))
), "../..");
const distDir = node_path.resolve(node_url.fileURLToPath(
  (typeof document === 'undefined' ? require('u' + 'rl').pathToFileURL(__filename).href : (_documentCurrentScript && _documentCurrentScript.src || new URL('preset.cjs', document.baseURI).href))
), "../..", "dist");
const runtimeDir = node_path.resolve(distDir, "runtime");
const pluginsDir = node_path.resolve(runtimeDir, "plugins");
const componentsDir = node_path.resolve(runtimeDir, "components");
const composablesDir = node_path.resolve(runtimeDir, "composables");
const dirs = [distDir, packageDir, pluginsDir, componentsDir, composablesDir];
let nuxt;
function extendComponents(nuxt2) {
  nuxt2.hook("components:extend", (components) => {
    const nuxtLink = components.find(({ name }) => name === "NuxtLink");
    nuxtLink.filePath = node_path.join(runtimeDir, "components/nuxt-link");
    nuxtLink.shortPath = node_path.join(runtimeDir, "components/nuxt-link");
    nuxt2.options.build.transpile.push(nuxtLink.filePath);
  });
}
async function extendComposables(nuxt2) {
  const { addImportsSources } = await import('@nuxt/kit');
  nuxt2.options.build.transpile.push(composablesDir);
  addImportsSources({ imports: ["useRouter"], from: node_path.join(composablesDir, "router") });
}
async function defineNuxtConfig(baseConfig) {
  const { loadNuxt, buildNuxt, addPlugin, extendPages } = await import('@nuxt/kit');
  nuxt = await loadNuxt({
    rootDir: baseConfig.root,
    ready: false,
    dev: false,
    overrides: {
      ssr: false
    }
  });
  if (nuxt.options.builder !== "@nuxt/vite-builder")
    throw new Error(`Storybook-Nuxt does not support '${nuxt.options.builder}' for now.`);
  let extendedConfig = {};
  nuxt.hook("modules:done", () => {
    extendComposables(nuxt);
    addPlugin({
      src: node_path.join(pluginsDir, "storybook"),
      mode: "client"
    });
    extendComponents(nuxt);
    extendPages((pages) => {
      pages.push({
        name: "storybook-iframe",
        path: "/iframe.html"
      });
    });
    nuxt.hook(
      "vite:extendConfig",
      (config, { isClient }) => {
        if (isClient)
          extendedConfig = vite.mergeConfig(config, baseConfig);
      }
    );
  });
  await nuxt.ready();
  try {
    await buildNuxt(nuxt);
    return {
      viteConfig: extendedConfig,
      nuxt
    };
  } catch (e) {
    throw new Error(e);
  }
}
const core = async (config) => {
  return {
    ...config,
    builder: "@storybook/builder-vite",
    renderer: "@storybook/vue3"
  };
};
const previewAnnotations = async (entry = []) => {
  return [...entry, node_path.resolve(packageDir, "preview")];
};
const viteFinal = async (config, options) => {
  const getStorybookViteConfig = async (c, o) => {
    const presetURL = node_url.pathToFileURL(node_path.join(await getPackageDir("@storybook/vue3-vite"), "preset.js"));
    const { viteFinal: ViteFile } = await import(presetURL.href);
    if (!ViteFile)
      throw new Error("ViteFile not found");
    return ViteFile(c, o);
  };
  const nuxtConfig = await defineNuxtConfig(await getStorybookViteConfig(config, options));
  return vite.mergeConfig(nuxtConfig.viteConfig, {
    build: { rollupOptions: { external: ["vue", "vue-demi"] } },
    define: {
      __NUXT__: JSON.stringify({ config: nuxtConfig.nuxt.options.runtimeConfig })
    },
    server: {
      fs: { allow: [vite.searchForWorkspaceRoot(process.cwd()), ...dirs] }
    },
    envPrefix: ["NUXT_"]
  });
};
async function getPackageDir(frameworkPackageName) {
  try {
    const require$1 = node_module.createRequire((typeof document === 'undefined' ? require('u' + 'rl').pathToFileURL(__filename).href : (_documentCurrentScript && _documentCurrentScript.src || new URL('preset.cjs', document.baseURI).href)));
    const packageDir2 = node_path.dirname(require$1.resolve(node_path.join(frameworkPackageName, "package.json"), { paths: [process.cwd()] }));
    return packageDir2;
  } catch (e) {
  }
  throw new Error(`Cannot find ${frameworkPackageName},`);
}

exports.core = core;
exports.previewAnnotations = previewAnnotations;
exports.viteFinal = viteFinal;
